### DISCUSSION TITLE
Can mock compile SRPMS in an order that doesn't result in a build error

### DISCUSSION THREAD

**Discussion by ashuk2000** [rpm-software-management/mock#1049] [Q&A] [answered]
if I want to compile lets say 3 source rpms and lets say there is some kinda dependency between them such that they can be only compiled in a certain order ( let's say srpm2--> srpm3--> srpm1) . Is there any way for mock to find this order first before compilation so that we can compile it in one shot. I have tried out the --recurse and --rebuild with --chain, but this way it seems like a brute force way.

**Comment by Conan-Kudo** [comment, marked as answer]
You have to specify the order. It processes SRPMs in the order you give them when you use `mock --chain`.

**Reply by ashuk2000** [reply]
But the info for dependency is in the spec files , can't we have a option  which will go through the specs and create graph/topo sort it and get the proper order to compile them.

**Comment by xsuchy** [comment]
For context: several people already had this idea. Me too. This must be easy, right... I will do this over the weekend... Hmm, what to do with the %if conditions? This must be evaluated in the target chroot. And what about Software collections where macros are evaluated in the buildroot? We have to at least initiate the buildroot. And what about Dynamic buildrequires? We have to reach the %prep phase... Well, the package is almost built already - the brute force method does not sound so bad, let's keep it.

**Comment by jafd** [comment]
Also there are packages depending on those `cmake(x)`, `python3dist(x)`, `perl(x)`, `pkgconfig(x)` which you cannot know how to get _before_ you build your sources, those `Provides:` lines get populated dynamically...

What I do is discover the dependencies the hard way first, then create a Makefile which specifies what depends on what. I write it once, then remaking builds things once and in a correct order. Well, since I've been doing it for quite a long time, the Makefile got a bit more involved, but this basic idea is still there.
